app/cdap/components/GridTable/index.tsx

```diff
@@ -231,88 +312,231 @@ export default function GridTable() {
     }
-    const rowData =
+    const rowData: IRowData[] =
       rawData &&
       rawData.values &&
-      Array.isArray(rawData.values) &&
-      rawData.values.map((eachRow) => {
-        const { body, ...rest } = eachRow;
+      Array.isArray(rawData?.values) &&
+      (rawData?.values.map((eachRow: IRecords) => {
+        const { ...rest } = eachRow;
         return rest;
-      });
-
+      }) as IRowData[]);
+    dispatch({
+      type: IGridTableActions.TABLE_META_INFO,
+      payload: {
+        columnCount: rawData?.headers?.length,
+        rowCount: rawData?.values?.length - 1,
+      },
+    });
     setRowsDataList(rowData);
   };
 
+  const showRecipePanelHandler = () => {
+    setShowRecipePanel((prev) => !prev);
+  };
+
+  useEffect(() => {
+    if (snackbarState.open) {
+      setTimeout(() => {
+        setSnackbar({
+          open: false,
+          isSuccess: false,
+          message: ``,
+        });
+      }, 5000);
+    }
+  }, [snackbarState]);
+
+  const applyDirectiveAPICall = (newDirective, action, removed_arr, from) => {
+    setLoading(true);
+    const { dataprep } = DataPrepStore.getState();
+    const { workspaceId, workspaceUri, directives, insights } = dataprep;
+    let gridParams = {};
+    const updatedDirectives = action === 'add' ? directives.concat(newDirective) : newDirective;
+    const requestBody = directiveRequestBodyCreator(updatedDirectives);
+    const arr = JSON.parse(JSON.stringify(newDirective));
+    requestBody.insights = insights;
+
+    const workspaceInfo = {
+      properties: insights,
+    };
+    gridParams = {
+      directives: updatedDirectives,
+      workspaceId,
+      workspaceUri,
+      workspaceInfo,
+      insights,
+    };
+    const payload = {
+      context: params.namespace,
+      workspaceId: params.wid,
+    };
+    MyDataPrepApi.execute(payload, requestBody).subscribe(
+      (response) => {
+        DataPrepStore.dispatch({
+          type: DataPrepActions.setWorkspace,
+          payload: {
+            data: response.values,
+            values: response.values,
+            headers: response.headers,
+            types: response.types,
+            ...gridParams,
+          },
+        });
+        setLoading(false);
+        setGridData(response);
+        setShowRecipePanel(false);
+        setSnackbar({
+          open: true,
+          isSuccess: true,
+          message:
+            action === 'add'
+              ? `Transformation ${arr} successfully added`
+              : from === 'undo' || arr?.length === 0
+              ? 'Transformation successfully deleted'
+              : `${removed_arr?.length} transformation successfully deleted from ${
+                  arr[arr.length - 1]
+                }`,
+        });
+      },
+      (err) => {
+        setSnackbar({
+          open: true,
+          isSuccess: false,
+          message: `Failed to transform ${newDirective}`,
+        });
+        setLoading(false);
```

Comment:
if you're going to create a reducer for this you might as well more most of the state logic to that reducer and de clutter this component

Reply:
Done.